src/client/cl_entity.c

```diff
@@ -348,38 +348,61 @@ static void Cl_UpdateLerp(void) {
+ * @remarks This ensures that each server frame is interpolated at least once, so that its effects
+ * are carried through to the view. Thus, the view is updated at each iteration below, but only 
+ * the output of the last iteration will be rendererd.
  */
 void Cl_Interpolate(void) {
 
-	Cl_UpdateLerp();
+	if (cl.render_frame == NULL) {
+		cl.render_frame = &cl.frames[cl.frame.frame_num & PACKET_MASK];
+	}
 
-	for (uint16_t i = 0; i < cl.frame.num_entities; i++) {
+	const int32_t last_frame_num = cl.frame.frame_num;
+	const int32_t frames = last_frame_num - cl.render_frame->frame_num;
+	if (frames > 1) {
+		Com_Debug("Running %d frames\n", frames);
+	}
 
-		const uint32_t snum = (cl.frame.entity_state + i) & ENTITY_STATE_MASK;
-		cl_entity_t *ent = &cl.entities[cl.entity_states[snum].number];
+	for (int32_t i = cl.render_frame->frame_num; i <= last_frame_num; i++) {
 
-		if (!VectorCompare(ent->prev.origin, ent->current.origin)) {
-			VectorLerp(ent->prev.origin, ent->current.origin, cl.lerp, ent->origin);
-			ent->lighting.state = LIGHTING_DIRTY;
-		} else {
-			VectorCopy(ent->current.origin, ent->origin);
-		}
+		cl.frame = cl.frames[i & PACKET_MASK];
+		cl.delta_frame = cl.frame.delta_frame_num <= 0 ? NULL : &cl.frames[cl.frame.delta_frame_num & PACKET_MASK];
 
-		if (!VectorCompare(ent->prev.termination, ent->current.termination)) {
-			VectorLerp(ent->prev.termination, ent->current.termination, cl.lerp, ent->termination);
-			ent->lighting.state = LIGHTING_DIRTY;
-		} else {
-			VectorCopy(ent->current.termination, ent->termination);
-		}
+		Cl_UpdateLerp();
 
-		if (!VectorCompare(ent->prev.angles, ent->current.angles)) {
-			AngleLerp(ent->prev.angles, ent->current.angles, cl.lerp, ent->angles);
-			ent->lighting.state = LIGHTING_DIRTY;
-		} else {
-			VectorCopy(ent->current.angles, ent->angles);
+		for (uint16_t j = 0; j < cl.frame.num_entities; j++) {
+
+			const uint32_t snum = (cl.frame.entity_state + j) & ENTITY_STATE_MASK;
+			cl_entity_t *ent = &cl.entities[cl.entity_states[snum].number];
+
+			if (!VectorCompare(ent->prev.origin, ent->current.origin)) {
+				VectorLerp(ent->prev.origin, ent->current.origin, cl.lerp, ent->origin);
+				ent->lighting.state = LIGHTING_DIRTY;
+			} else {
+				VectorCopy(ent->current.origin, ent->origin);
+			}
+
+			if (!VectorCompare(ent->prev.termination, ent->current.termination)) {
+				VectorLerp(ent->prev.termination, ent->current.termination, cl.lerp, ent->termination);
+				ent->lighting.state = LIGHTING_DIRTY;
+			} else {
+				VectorCopy(ent->current.termination, ent->termination);
+			}
+
+			if (!VectorCompare(ent->prev.angles, ent->current.angles)) {
+				AngleLerp(ent->prev.angles, ent->current.angles, cl.lerp, ent->angles);
+				ent->lighting.state = LIGHTING_DIRTY;
+			} else {
+				VectorCopy(ent->current.angles, ent->angles);
+			}
 		}
+
+		Cl_UpdateView();
 	}
+
+	cl.render_frame = &cl.frames[cl.frame.frame_num & PACKET_MASK];
 }
 
```

Comment:
This is the meat of the change. Basically, after reading all updates from the server, enter a loop where we start at the last frame we simulated, and run up to the most recently received server frame. At each iteration, we fully populate the view. This ensures nothing falls through the cracks, so to speak.